src/psyclone/tests/psyir/frontend/fparser2_interface_block_test.py

```diff
@@ -0,0 +1,154 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' Performs py.test tests on the handling of interface blocks in the fparser2
+    PSyIR front-end. '''
+
+from __future__ import absolute_import
+
+from fparser.two import Fortran2003
+from psyclone.psyir.nodes import Container, CodeBlock, FileContainer
+from psyclone.psyir.symbols import RoutineSymbol, Symbol, UnknownFortranType
+
+
+def test_named_interface(fortran_reader):
+    ''' Test that the frontend creates a RoutineSymbol of UnknownFortranType
+    for a named interface block.'''
+    dummy_module = '''
+    module dummy_mod
+      private
+      public eos
+      interface eos
+        module procedure eos_insitu, eos_insitu_2d
+      end interface
+    contains
+      subroutine eos_insitu(f1, f2)
+          real, intent(in)  :: f1
+          real, intent(out) :: f2
+          f2 = f1 + 1
+      end subroutine eos_insitu
+      subroutine eos_insitu_2d(f1, f2)
+          real, dimension(:,:), intent(in)  :: f1
+          real, dimension(:,:), intent(out) :: f2
+          f2(:,:) = f1(:,:) + 1
+      end subroutine eos_insitu_2d
+    end module dummy_mod
+    '''
+    file_container = fortran_reader.psyir_from_source(dummy_module)
+    container = file_container.children[0]
+    assert isinstance(container, Container)
+    assert container.symbol_table.lookup("eos_insitu")
+    eos = container.symbol_table.lookup("eos")
+    assert isinstance(eos, RoutineSymbol)
+    assert isinstance(eos.datatype, UnknownFortranType)
+    assert (eos.datatype.declaration == "interface eos\n"
+            "  module procedure eos_insitu, eos_insitu_2d\n"
+            "end interface")
+    assert eos.visibility == Symbol.Visibility.PUBLIC
+
+
+def test_generic_interface(fortran_reader):
+    ''' Test that we create a CodeBlock if a module contains an interface block
+        that defines interfaces to external routines. '''
+    dummy_module = '''
+    module dummy_mod
+      INTERFACE
+      SUBROUTINE EXT1 (X, Y, Z)
+      REAL, DIMENSION (100, 100) :: X, Y, Z
+      END SUBROUTINE EXT1
+      SUBROUTINE EXT2 (X, Z)
+      REAL X
+      COMPLEX (KIND = 4) Z (2000)
+      END SUBROUTINE EXT2
+      FUNCTION EXT3 (P, Q)
+      LOGICAL EXT3
+      INTEGER P (1000)
+      LOGICAL Q (1000)
+      END FUNCTION EXT3
+      END INTERFACE
+    end module dummy_mod
+    '''
+    container = fortran_reader.psyir_from_source(dummy_module)
+    assert isinstance(container, FileContainer)
+    assert len(container.children) == 1
+    assert isinstance(container.children[0], CodeBlock)
```

Comment:
This could still be problematic as although the output will contain the Interface, the symbol table will not contain EXT1, EXT2 and EXT3, and it may refer to them, is that right? No need to be solved it in this PR but maybe worth mentioning it.

Reply:
You had me worried there but actually, it's OK because we now get a CodeBlock for the whole module containing such an interface definition. I've added a comment.

Reply:
I know it creates a CodeBlock and kind of solve the problem is they aren't touched, my worry was for EXT1, EXT2 and EXT3 symbols not being declared and not accessible from inside the module "contains".

Reply:
I don't quite follow - the CodeBlock will contain the whole module - both any variable/interface declarations and any routines within a CONTAINS section? Do you mean if some other code `use`s the module then we won't be able to find (any) variables that it defines?